pkg/cloudprovider/provider/vsphere/helper.go

```diff
@@ -360,3 +412,18 @@ func validateDiskResizing(disks []*types.VirtualDisk, requestedSize int64) error
 	}
 	return nil
 }
+
+//getDatastoreFromVM gets the datastore where the VM files are located.
+func getDatastoreFromVM(ctx context.Context, session *Session, vmRef *object.VirtualMachine) (*object.Datastore, error) {
```

Comment:
You create a vm based on specific configurations, such as whether this vm is using a datastore cluster or datastore. And based on your implementation you made it distinct by using different names in the template: `datastore` or `datastoreCluster`. Those two properties are mutually exclusive which means they will never be set for the same machines set. And since you have all this information in the config a simple if statement can decide what is the right path for this machine:

```
if config.DatastoreCluster != "" && config.Datastore == "" {
  return session.Finder.Datastore(ctx, config.DatastoreCluster)
} else {
  return session.Finder.Datastore(ctx, config.Datastore)
}
```
The places where this function is needed have already access to that information.

Reply:
this does not work: `session.Finder.Datastore(ctx, config.DatastoreCluster)`.  It would mean trying to lookup a datastore with the datastore cluster name we had in input.

Reply:
`*object.Datastore` and `*object.StoragePod` are different, so I think it's not so straightforward. Even `resolveDatastoreRef` returns a specific datastore based on a cluster, except it is much more complicated than `getDatastoreFromVM`, so maybe just looking up the specific datastore where the VM already is, is not such a bad idea.